Teams.Security/CurrentUser.cs

```diff
@@ -0,0 +1,39 @@
+﻿using Microsoft.AspNetCore.Http;
+using System.Security.Authentication;
+using System.Security.Claims;
+
+namespace Teams.Security
+{
+    public class CurrentUser : ICurrentUser
```

Comment:
You can use Lazy approach here and it will looks like _accessor = new Lazy<T>(() => LoadUserDetails()).  In LoadUserDetails you can intercept all required conditions